### DISCUSSION TITLE
Outreachy Internship: Development of Zubhubs Mobile application

### DISCUSSION THREAD

**Discussion by alicendeh** [unstructuredstudio/zubhub#486] [Ideas]
Below are answers based on questions are by @srish and @tuxology on the Zulip chat:

1. **Where will the code live**
1.1 We can either create a folder in the existing project where the code will reside or
1.2 We create a repo for the mobile app under the unstructured studio org and the mobile app exists as a standalone repo(I think this is best).

2. **What web features are doable for the mobile application for a 3 month project? Let's make a priority list.**
2.1 Authentication 
2.1 (Home page) which includes fetching all projects and displaying to the user(Same as on the web app)
2.3 Dashboard module which includes 
    - View currently logged in user profile
    -  Edit/delete account
    -  View all projects

3. Project creation
4. Commenting
5. Following/Unfollowing creators/Likes and unlike/bookmarks
6. Translation

**Comment by alicendeh** [comment]
**REFINING THE ABOVE ANSWERS**

The very first module is the **Authentication module**.
The Authentication module is divided into two sub-modules(**LOGIN** and **SIGNUP**)

**SUB MODULE 1: SIGN IN**
- From the web app, the signin has the following fields(email, password).
- We can continue using email and password to login users. This is because the current web app uses this means and a lot of users are used to login in through this way already.
- I personally think we don't need to track users who are signing in from the web app or from the mobile app unless we will need that data for analytics purposes.
- Since it's a mobile app, we don't need to log out the users when they close the app. Rather whenever the users come to the app, we check if it's their first time or not. If it's their first time, we prompt them to log in else we route them to the login screen.
- For backend-oriented modifications, The API will need to grant access to the IP from where the mobile app is emitted so as to enable requests to be made.

**SUB MODULE 2: SIGN UP**
 From the web app, the sign-in has the following fields(username, location, dob, phone, email, password, confirm password, and bio)
 - For the location, I suggest we use 
1. react-native-country-picker-modal. This package provides a list of countries, their flags, and country code
2. React Native Phone Number Input: This will help us in managing the phone number formatting and ensuring the number is valid based on the country
3. After sending the data to the backend server, we need to handle all cases of error(incorrect password, empty fields and all. The rest.)
4. We can use a toast to display all errors returned (As on the web app)

**Reply by tuxology** [reply]
This is good progress. Things are getting clearer now! I also suggest you create tiny sketches (can be hand drawn/rough - doesn't matter) and annotate them with individual points when required. You can refer some work done by previous interns on a very similar project in their planning phase: https://meta.wikimedia.org/wiki/WikiContrib/Proposed_implementation General suggestion IMO, is that we need to be almost 70% sure of what we are building (**requirement specification**) and how it may look (**mockups**) and how it would function (**functional specification**). Only then can/should we begin actual coding.

A few comments:

> I personally think we don't need to track users who are signing in from the web app or from the mobile app unless we will need that data for analytics purposes.

OK, fair enough, we can skip this for now. However, we should have some way in backend to identify the incoming request may want some mobile specific info. We may modify APIs as and when required I would say.

> For backend-oriented modifications, The API will need to grant access to the IP from where the mobile app is emitted so as to enable requests to be made.

I didn't understand this one. Can you elaborate? Rest all looks quite nice :+1:

**Comment by alicendeh** [comment]
**DEEP DIVE INTO THE AUTHENTICATION MODULE**

**1. AUTHENTICATION MODULE MOCKUPS**
I did some sketches on how the Authentication module will look like

**A. LOGIN**
![IMG_1371](https://user-images.githubusercontent.com/58989178/172209699-48f95687-fb4c-4567-acbe-d4a1363d8c79.jpg)

**B. REGISTER**
![WhatsApp Image 2022-06-06 at 6 18 12 PM](https://user-images.githubusercontent.com/58989178/172212366-c5eb1f1d-d6f3-427f-b2ba-9c9d6323718e.jpeg)

**C. FORGET PASSWORD**
![WhatsApp Image 2022-06-06 at 6 18 19 PM](https://user-images.githubusercontent.com/58989178/172212398-dc5a1ea7-6783-4e91-95dd-1fa3066b6563.jpeg)

**D. BOTTOM TAB**
 My idea is, when the user comes to the app for the first time we have a bottom tab bar with two tabs
i. The home tab were we have all projects
ii. Authentication tab were the users can either login or signup
![WhatsApp Image 2022-06-06 at 6 18 19 PM (1)](https://user-images.githubusercontent.com/58989178/172213507-9b781818-0c6f-47c9-ad36-3d001046142b.jpeg)


**2. END POINTS AND DATA NEEDED**
 
**A. LOGIN**
**ENDPOINT:** /api/rest-auth/login/
**Required data:** Username and Password
  
**B. SIGNUP**
**ENDPOINT:** /api/creators/register/
**Required data:**  
     username,
      email,
      phone,
      dateOfBirth,
      location,
      password1,
      password2,
      bio,
      subscribe,

**C. FORGET PASSWORD**
To be added..

**Reply by tuxology** [reply]
@alicendeh Good ideas, well organized! :+1: Just a few comments. Rest all looks good so far!
 - So we had some good ideas here : https://github.com/unstructuredstudio/zubhub/discussions/348#discussioncomment-2546024 We discussed the behavior of an unauthenticated user visiting the app for the first time. In that case, maybe the bottom bar should have a Home Button + search button + Add Project? Clicking home/search takes you to "Projects" view (which you will eventually mockup I presume) and Clicking "Add Project" takes you to "Login" or "Sign Up"
 - The reason is that we want users not authenticated to still browse Zubhub and projects, but we periodically nudge them to sign up. We can do this by 
   - Showing them a sliding down element on top of the app which persists after being shown the first time until login/signup is complete. When they click that, they go through the auth workflow you sketched
   - We also take them to auth workflow anytime they try to leave a comment, clap, bookmark etc. 
   - The top right corner circle shows a generic human image. Clicking that open up login/signup workflow
   - Send Push notifications from time to time. We just need to be make sure we don't spam. Just slight nudges every few days.

**Reply by alicendeh** [reply]
@tuxology I have a suggestion to make.Can we do it in a way that when the users arent logged in yet, we dont show them a bottom tab bar. Rather we show to them a top tab bar containing an avater,the search and language toggler and when they click on the avater,we bring them to the authentication flow.Example here.
<img width="362" alt="Screen Shot 2022-06-09 at 07 36 14" src="https://user-images.githubusercontent.com/58989178/172780772-1dd2b31d-9479-41a9-afbe-e989c4e481a0.png">

And when they login,we can then introduce the bottom tab bar

**Reply by alicendeh** [reply]
And also i want to ask which is prefferable. Should we leave the white background as it is above or we should go witha yellowish background.As shouwn below
<img width="342" alt="Screen Shot 2022-06-09 at 07 39 34" src="https://user-images.githubusercontent.com/58989178/172781266-599157e4-8d63-4ccf-988c-1c8b7be2611a.png">

**Reply by tuxology** [reply]
I am inclined to suggest a gradient yellow view like we have on our current responsive mobile web view

**Comment by alicendeh** [comment]
Hello @tuxology and @srish here is the link to the design file containing the Authentication flow.(Without the bottom bar) https://www.figma.com/file/yRY88ub5V8F20JWY7dTvMl/ZUBHUB_MOBILE_APP?node-id=0%3A1

**Reply by kamthamc** [reply]
I have some feedback 
* Regarding language selection menu, users won't change it frequently. I think only frequent items should be quickly accessible. So move it to profile or hamburger menu?
* The buttons color should be same red color or the primary color the host chose. 
* Two text fields on same line in a mobile screen might not be a good idea?

**Reply by alicendeh** [reply]
Thanks for the feedback @kamthamc. 
1. For the language, we can take the device language and if the users want to change it, they can then do that in their settings. Very good suggestion thanks
2. I buy the idea that the color of the button should be read as well. I gave them the teal color because that was what I saw on the current web app. @tuxology please what do you think about this?
3. I will rectify that.
Thanks a lot for the feedback @kamthamc

**Reply by tuxology** [reply]
@alicendeh These are some great mockups! :clap:  I specially like the Sign Up nudge from below! I've left some comments in the mockups. Nothing major except just one thing - we can think about creating a wizard style navigation for account creation instead of just a single page form? I don't have any strong preference - feel free to use what feels best for you.

I've left comments on some individual components in the UI.

> I buy the idea that the color of the button should be read as well. I gave them the teal color because that was what I saw on the current web app. @tuxology please what do you think about this?

I think lets keep the teal color for now for visual consistency!

**Reply by alicendeh** [reply]
Thanks for the feedback @tuxology I will do the necesary modifications and work on the next feature
